core/store/src/trie/mem/loading.rs

```diff
@@ -75,23 +87,118 @@ pub fn load_trie_from_flat_state(
     Ok(tries)
 }
 
+fn get_state_root(
+    store: &Store,
+    block_hash: CryptoHash,
+    shard_uid: ShardUId,
+) -> Result<CryptoHash, StorageError> {
+    let chunk_extra = store
+        .get_ser::<ChunkExtra>(DBCol::ChunkExtra, &get_block_shard_uid(&block_hash, &shard_uid))
+        .map_err(|err| {
+            StorageError::StorageInconsistentState(format!(
+                "Cannot fetch ChunkExtra for block {} in shard {}: {:?}",
+                block_hash, shard_uid, err
+            ))
+        })?
+        .ok_or_else(|| {
+            StorageError::StorageInconsistentState(format!(
+                "No ChunkExtra for block {} in shard {}",
+                block_hash, shard_uid
+            ))
+        })?;
+    Ok(*chunk_extra.state_root())
+}
+
+/// Constructs in-memory tries for the given shard, so that they represent the
+/// same information as the flat storage, including the final state and the
+/// deltas. The returned tries would contain a root for each block that the
+/// flat storage currently has, i.e. one for the final block, and one for each
+/// block that flat storage has a delta for, possibly in more than one fork.
+pub fn load_trie_from_flat_state_and_delta(
+    store: &Store,
+    shard_uid: ShardUId,
+    maximum_arena_size: usize,
+) -> Result<MemTries, StorageError> {
+    debug!(target: "memtrie", "Loading base trie for {} from flat state...", shard_uid);
```

Comment:
Stealing nit from parallel PR: https://github.com/near/nearcore/pull/10107#discussion_r1383451722

Consider more structured approach to logging:
```
debug!(target: "memtrie", %shard_uid, "Loading base trie from flat state...");
...
debug!(target: "memtrie", %shard_uid, sorted_deltas_len = sorted_deltas.len(), "Ready to apply deltas");
...
```